core/api/src/services/ledger/paginated-ledger.ts

```diff
@@ -46,27 +46,59 @@ export const paginatedLedger = async ({
     filterQuery["_id"] = { $gt: new Types.ObjectId(before) }
   }
 
-  let limit = first ?? MAX_PAGINATION_PAGE_SIZE
-  let skip = 0
+  if (filters.username) {
+    filterQuery["username"] = filters.username
+  }
 
-  const total = await Transaction.countDocuments(filterQuery)
+  if (filters.addresses) {
+    filterQuery["payee_addresses"] = { $in: filters.addresses }
+  }
+
+  const documentCount = await Transaction.countDocuments(filterQuery)
+
+  // hasPreviousPage and hasNextPage can default to false for the opposite pagination direction per the Connection spec
+  let hasPreviousPage = false
+  let hasNextPage = false
+  let transactionRecords: ILedgerTransaction[] = []
 
-  if (last) {
-    limit = last
-    if (total > last) {
-      skip = total - last
+  if (first !== undefined) {
+    if (documentCount > first) {
+      hasNextPage = true
     }
+
+    transactionRecords = await Transaction.collection
+      .find<ILedgerTransaction>(filterQuery)
+      .sort({ _id: -1 })
+      .limit(first)
+      .toArray()
+  } else {
+    let skipAmount = 0
+    if (documentCount > last) {
+      hasPreviousPage = true
+      skipAmount = documentCount - last
+    }
+
+    transactionRecords = await Transaction.collection
+      .find<ILedgerTransaction>(filterQuery)
+      .sort({ _id: -1 })
+      .skip(skipAmount)
+      .toArray()
   }
 
-  const slice = await Transaction.collection
-    .find<ILedgerTransaction>(filterQuery)
-    .sort({ datetime: -1, timestamp: -1, _id: -1 })
-    .limit(limit)
-    .skip(skip)
-    .toArray()
+  const txs = transactionRecords.map((tx) => translateToLedgerTx(tx))
 
   return {
-    slice,
-    total,
+    edges: txs.map((tx) => ({
+      cursor: checkedToPaginatedQueryCursor(tx.id),
+      node: tx,
+    })),
+    pageInfo: {
+      startCursor: txs[0]?.id ? checkedToPaginatedQueryCursor(txs[0].id) : undefined,
+      endCursor: txs[txs.length - 1]?.id
+        ? checkedToPaginatedQueryCursor(txs[txs.length - 1].id)
```

Comment:
same question here

Reply:
Then the end cursor will be undefined, which is the expected behavior.